### ISSUE TITLE
Implement Lifecycle Configuration

### ISSUE THREAD

**Issue by afaenger** [minio/minio#5139]
It would be great if minio added support for bucket lifecycle configurations, e.g. to perform automatic object expiration after some days.

**Comment by nitisht** [comment]
@afaenger you can use minio-client (mc) for that..

For example, to delete an object older than 7 days

```sh
mc rm --older-than 7 myminio/testbucket/object
```

**Comment by nitisht** [comment]
@afaenger feel free to close the issue if there is nothing else to be addressed.

**Comment by nitisht** [comment]
Closing due to no response.

**Comment by SeanFromIT** [comment]
This should be a server-side activity. In an enterprise environment, you cannot trust users to do their own cleanup. You're lucky if they set the lifecycle configuration.

**Comment by harshavardhana** [comment]
It was intentional to keep it client side. It is not users it is admin who runs mc and cleans up periodically as needed.

**Comment by SeanFromIT** [comment]
Correct, users should set the policy. But this isn't a one-time activity and as such should be automated.

**Comment by afaenger** [comment]
That's right; we have many applications that use their own buckets, it's impossible to setup a task for each application on the server to cleanup the expired files whereas a lifecycle configuration is easy to define per application.

**Comment by harshavardhana** [comment]
Client side means more control and visibility. Doing it on server side means complexity and makes server a black box.

We wish to keep server simple and avoid unnecessary APIs which can be cleanly more easily done on client side with more control.

**Comment by afaenger** [comment]
But then minio cannot be used easily as a drop in replacement for Amazon S3 or other implementations as one has to manually take care of deleting objects; in our case setting the lifecycle configuration is part of the application and fails as minio does not support this api/functionality.

**Comment by dewjam** [comment]
I'm in agreement with @SeanFromIT .  The bucket lifecycle policies should be set by the client, but the expiration and removal of objects should be handled on the server side (automatically).  Seems like a pretty common use-case that would make Minio more appealing to enterprises.  Can we reopen this issue?  Get some more eyes on it?

**Comment by lucasoares** [comment]
Repoen this issue, please!

**Comment by harshavardhana** [comment]
Minio server code would like to be simpler and predictable when dealing with the user I/O. Not mangle with the actual namespace etc. Giving more control to the user using `mc` provides visibility and control over this behavior so the previous suggestions still stand. 

One can simply set up an `mc` in a cron-job on the server side to do this periodically whenever you don't want it just turn it off if need to expire faster simply change the number of days - all of this can be done without making any API calls whatsoever if all you require is simply just to delete some expired objects. This is how we periodically purge objects on our public server https://play.minio.io:9000

**Comment by dewjam** [comment]
@harshavardhana The solution you suggested is certainly feasible and acceptable in many situations.  However, I don't think it would fly very far in an Enterprise.  

Of course it depends on the intended use-case for Minio as a whole.  Is it targeted at SMBs and developers who just need a simple S3 interface?  Or is Minio intended to be a feasible solution for the enterprise?  

Based on the features that are already available and how Minio is being positioned in the market, I would think the latter is true.  Features like encryption, erasure coding, and high availability are totally enterprise focused.  And the semi-recent funding round would point to a broader use-case as well (I think it was referred to as a "universal data-access layer").

I do believe this is a feature already present in the Amazon S3 api spec as well.
https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lifecycle-mgmt.html

**Comment by harshavardhana** [comment]
> I do believe this is a feature already present in the Amazon S3 API spec as well.
> https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lifecycle-mgmt.html

We are certainly aware that it is present, there are many other APIs that are present as well. We don't implement ACLs as well since we think those are legacy and bucket policies should be used. The decision here is quite similar. We can justify pretty much everything on why such and such API AWS S3 implements and why Minio should do that too.

Lifecycle is more of a bucket management API, so without disturbing Minio, our suggestion is this can be done by using `mc`. 

> The solution you suggested is certainly feasible and acceptable in many situations. However, I don't think it would fly very far in an Enterprise.

When it is acceptable in many situations it is quite simply the right solution as well.  Minio is indeed enterprise focussed, most of our community is enterprise anyways. So we do appreciate great feedback like this.

**Comment by dewjam** [comment]
>  The solution you suggested is certainly feasible and acceptable in many situations. However, I don't think it would fly very far in an Enterprise.

My point with this comment is that the mentioned approach is not acceptable for most Enterprises.  And if Enterprise is the focus, I would expect this to at least be on the roadmap.  (Definately not expecting it to be a top priority)

**Comment by atrauzzi** [comment]
Anything to do with expiry - despite seeming like a black-box feature - seems like it ought to be in scope for this project.  I definitely agree that there's a component of _magic_ here, but this is a very common feature that people bank on.

**Comment by IAkumaI** [comment]
Hm. So, how can i use `mc` to cleanup bucket while a minio running inside a docker container?
I don't want to configure cron on a server because...I use docker.

**Comment by harshavardhana** [comment]
> Hm. So, how can i use mc to cleanup bucket while a minio running inside a docker container?
> I don't want to configure cron on a server because...I use docker.

you can use `docker pull minio/mc`

**Comment by IAkumaI** [comment]
Yeah, but...imagine I use swarm. I dont even know which physical server right now runs minio service. So how I should configure cron? On each server?
And configure for what? I dont even know minio's container name and it can have different names on different servers.

**Comment by dewjam** [comment]
@IAkumaI I think you would run the `minio/mc` container independent of the "minio/minio" (server) containers.  In the `minio/mc` container, you might run a basic script that connects to the minio server and cleans up old objects on a regular basis (using the `mc` command).  This is a trivial task on it's own, but the fact remains that it will become unmanageable at scale.  When you have many buckets, tenants, etc.

@harshavardhana I do understand this feature is not one you're fond of implementing in the Minio code base.  However, it seems like the community has expressed the importance of this capability.  Is there any possibility it could be at least added to the road-map?

(is there a publicly available feature road-map, by the way?)

**Comment by harshavardhana** [comment]
We are planning to implement this [API](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUTlifecycle.html) - thanks for all the feedback till now to everyone involved, we will post a proposal once the design is finalized here.

**Comment by rubenfonseca** [comment]
Thanks for all your nice work! Any idea when lifecycle policies will get implemented on minio?

**Comment by harshavardhana** [comment]
> Thanks for all your nice work! Any idea when lifecycle policies will get implemented on minio?

Soon @rubenfonseca, unfortunately, don't have a set date at the moment.

**Comment by mthiyagarajanveritone** [comment]
Thanks to minio team for all their great work.. When i searched for setting TTL at a bucket level i stumbled upon this issue. Whether lifecycle policies has been implemented for minio buckets?

**Comment by nikwen** [comment]
@mthiyagarajanveritone They haven't.

**Comment by npfistner** [comment]
Any news concerning lifecycle support?

I tried using S3 SDK setBucketLifecycleConfiguration against a loacal minio:minio/latest docker instance and ran into 
`com.amazonaws.services.s3.model.AmazonS3Exception: A header you provided implies functionality that is not implemented (Service: Amazon S3; Status Code: 501; Error Code: NotImplemented; ...`

**Comment by harshavardhana** [comment]
> Any news concerning lifecycle support?

@npfistner we have deferred this to future for now.

**Comment by npfistner** [comment]
Same with minio:minio/edge:
`com.amazonaws.services.s3.model.AmazonS3Exception: A header you provided implies functionality that is not implemented (Service: Amazon S3; Status Code: 501; Error Code: NotImplemented;`
Using AWS SDK com.amazonaws:aws-java-sdk-bom:1.11.531

**Comment by npfistner** [comment]
What a pitty. Bucket Lifecycle configuration support would make it a lot easier to use minio for us.

**Comment by npfistner** [comment]
Any suggestions for a workaround?

**Comment by ahilsend** [comment]
> Any suggestions for a workaround?

We currently have a cronjob with `mc rm --older-than 29 --recursive --force minio/<bucket>` to cleanup content older than 29 days. It would be better to have this built-in, but it does the job for now

**Comment by npfistner** [comment]
Mmh, we do have a single deadline for our objects. Depnding on tags we should be able to apply different expiration day limits. Seems not to be able with the mentioned minio client command `mc rm `

**Comment by npfistner** [comment]
An a cron job encreases deployment and management complexity. Seem not to fit really well int a dockerized enterprise environment.

**Comment by hamid-elaosta** [comment]
Minio could really benefit from a TTL per-object.

I have requirements to clean up stored objects after arbitrary periods.

This could be done with some meta and a clean job but a Redis-like TTL would be much easier to work with.

Other projects which normally use S3 are dependent on the lifecycle feature of AWS and Minio is therefore not "S3 compatible" in that sense.

**Comment by npfistner** [comment]
Now I'm confused:
The [Minio Java API Client reference](https://docs.min.io/docs/java-client-api-reference#setBucketLifeCycle) states support for the feature. 
Is this feature really working with the minio server?

**Comment by harshavardhana** [comment]
> The [Minio Java API Client reference](https://docs.min.io/docs/java-client-api-reference#setBucketLifeCycle) states support for the feature.

It was added as part of AWS S3 API support as a precursor for server support, but server change has taken lower priority at this point in time.

**Comment by npfistner** [comment]
Understood, but not happy about it ...

**Comment by lagorsse** [comment]
I would need that too

**Comment by mritd** [comment]
i need too... 😌

**Comment by oskapt** [comment]
There are situations where `mc` in a cron job isn't the right answer. For example, Sonatype Nexus Repository Manager can use S3 storage for local and proxied Docker registries. It has a cleanup policy based on publish date or date of last access. Neither of these are visible to `mc`. 

I might have downloaded an image a month ago, but my Kubernetes environment has used it within the last day. If I run `mc` and tell it to delete all blobs older than a week, then I've clobbered an image I should have kept.

Implementing lifecycle management is no doubt a significant undertaking. If it's in the roadmap, I hope that it finds its way into active development soon.

**Comment by kannappanr** [comment]
@oskapt We will start working on this feature soon.

**Comment by oskapt** [comment]
> @oskapt We will start working on this feature soon.

That's great news. Thank you @kannappanr. 

I want to thank all of you for Minio. I work for Rancher Labs (rancher.com), so I know how hard open source software can be, trying to balance the needs/wants of the community, the philosophical forces driving you to build something with no guarantee of reward, and the ever-present demands of real world economics.

Because of the work I do, I run a large lab in my home, with a 5-node Proxmox cluster, Kubernetes nodes, Synology boxes providing iSCSI and NFS storage, and because of where I live, I have a cap on bandwidth. Most of the services that use cloud resources operate as though everyone has unlimited bandwidth and budget. The idea that we can all just connect to S3 because it exists is as frustrating as all of my iOS devices pulling down the same 1.5 GB iOS update in the same night and then doing it again the next night for a patch release.

I have systems in place to cache, proxy, and block, but where I need a solution like S3, software like Minio truly saves the day. Being able to run an object store in a container on the NAS, and not having to dedicate RAM and CPU to a solution like Ceph (or Rook) makes it possible for me to make the most of my infrastructure.

I imagine that it seems like everyone just cries and demands work from you. Please know that those of us who use your software every day appreciate you **_every single day._**

**Comment by oskapt** [comment]
For reference, [here's the Nexus issue](https://issues.sonatype.org/browse/NEXUS-19052) that references Minio and asks for a hard delete option for blob stores using Minio for S3.

**Comment by einsty** [comment]
Your timing is impeccable!  I was literally working through the approach of using Nexus with an 8 node Minio cluster we have running on containers across ProxMox hosts as well!  This is fantastic @kannappanr and yes, as @oskapt  mentions... Thank you.

**Comment by npfistner** [comment]
Great news! Can't wait to drop our feature flag which cut of S3 lifycycle calls when using minio.
Thanks in advance!

**Comment by hamid-elaosta** [comment]
I've been looking at how I can resolve this issue in a project I'm working on for the short term. I had hoped I could use user metadata on objects to set a TTL and have a routine to iterate over the object list periodically and clean up expired ones.

Unfortunately, it seems that the user metadata cannot be used this way. Is my best option really going to be to track my object expiry in another database and remove them that way?

Has anyone else solved a similar issue without storing object meta/TTL data outside of Minio?

**Comment by krisis** [comment]
We will be adding object lifecycle support in phases. Initially, we will add support for object expiration matching object prefixes only. We will bring in support for object tagging and lifecycle support for tags subsequently. https://github.com/minio/minio/pull/7563 is the first in a series of PRs.

**Comment by wflanagan** [comment]
At least in my industry, we use s3 as an object cache.. when we write it, we want to set an expiration of something like 24 hours.. after which we would re-retrieve the item (and store in Minio).  We just implemented Minio in the last week, as an attempt to move off of S3 for doing this.  Was disappointed to see that per object TTL wasn't supported, but happy to see it's coming.

**Comment by lukepighetti** [comment]
Requesting an update on the status of this feature

**Comment by lukepighetti** [comment]
Glad to see this is being added

**Comment by PrathameshUpadhyay** [comment]
I see the PR is merged to master.
Can we get a release-doc, how to use this feature?

https://github.com/minio/minio/pull/7563

**Comment by vadmeste** [comment]
@PrathameshUpadhyay, that code is not fully functional yet, this is the PR which adds the expiry feature https://github.com/minio/minio/pull/7834 (transition is not there yet)

**Comment by harshavardhana** [comment]
Once #7834 is merged we will add a doc @PrathameshUpadhyay

**Comment by harshavardhana** [comment]
This feature is merged into master and with documentation https://github.com/minio/minio/tree/master/docs/lifecycle

Will be part of released version next week. Closing this as fixed..
